### PR TITLE
🏗♻🧽 Squeeze some bytes out of core

### PR DESCRIPTION
**Pull request by alanorozco**
- Use direct references where possible.
- `Object.defineProperty` helper to reduce redundancy.
- Leverage `indexOf` to use number-based `Layout` enum internally (thanks @sparhami for the idea).
- Compress `document-register-element` polyfill initialization table so that elements that follow the convention `'HTML$MyEl$Element': ['$myel$']` are generated on runtime from a flat array.

### CODE REVIEW

**Comment by alanorozco** [comment]
@choumx PTAL.

File: build-system/tasks/update-packages-document-register-element.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+/**
+ * Takes a document-register-element polyfill init object (like: git.io/fhxbR)
+ * in the form:
+ *
+ *  {
+ *    'collections': {
+ *      // ...
+ *    },
+ *    'elements': {
```
Comment (jridgewell) [inline build-system/tasks/update-packages-document-register-element.js]: Last I looked, the entire init object wasn't actually necessary for the DRE polyfill to work, it was a left over from some previous code. Either way, this is very risky change, and it won't be necessary with CEv1 polyfill.

File: extensions/amp-list/0.1/amp-list.js [hunk 2]
```diff
@@ -758,45 +758,43 @@ export class AmpList extends AMP.BaseElement {
...
-          width: '',
-        });
-        break;
-      case Layout.INTRINSIC:
-        this.element.classList.remove('i-amphtml-layout-intrinsic');
-        break;
-      default:
-        // fall through, always remove sizer and size-defined class
+    const {element} = this;
+    if (previousLayout == Layout.RESPONSIVE) {
```
Comment (jridgewell) [inline extensions/amp-list/0.1/amp-list.js]: It seems like there were fallthrough wins we could do in the switch statement.

File: src/layout.js [hunk 3]
```diff
@@ -44,19 +44,20 @@ export const LAYOUTS = [
...
-  FIXED_HEIGHT: 'fixed-height',
-  RESPONSIVE: 'responsive',
-  CONTAINER: 'container',
-  FILL: 'fill',
-  FLEX_ITEM: 'flex-item',
-  FLUID: 'fluid',
-  INTRINSIC: 'intrinsic',
+  // Values should match indices of `LAYOUTS`.
+  // Naming convention and indices enforced via tests.
+  NODISPLAY: LAYOUTS.indexOf('nodisplay'),
```
Comment (jridgewell) [inline src/layout.js]: Why is this better?

File: build-system/tasks/update-packages-document-register-element.js (hunk 1 above)
Comment (alanorozco) [reply build-system/tasks/update-packages-document-register-element.js]: 1. What's the status of CEv1?
2. How exactly is it risky?

File: extensions/amp-list/0.1/amp-list.js (hunk 2 above)
Comment (alanorozco) [reply extensions/amp-list/0.1/amp-list.js]: Will revert to switch.

File: src/layout.js (hunk 3 above)
Comment (alanorozco) [reply src/layout.js]: Saves a few bytes by having the enum keys be number literals instead of string literals.

**Comment by alanorozco** [comment]
Since the changes to custom-element are *controversial* and the changes to `Layout` enum only save like 70B I'll close this for now.

**Pull request closed without merging**